Account/FxAPushMessageHandler.swift

```diff
@@ -0,0 +1,220 @@
+            return deferMaybe(PushMessageError.notDecrypted)
+        }
+
+        guard let encoding = userInfo["con"] as? String, // content-encoding
+            let payload = userInfo["body"] as? String else {
+                return deferMaybe(PushMessageError.messageIncomplete)
+        }
+        // ver == endpointURL path, chid == channel id, aps == alert text and content_available.
+
+        let plaintext: String?
+        if let cryptoKeyHeader = userInfo["cryptokey"] as? String,  // crypto-key
+            let encryptionHeader = userInfo["enc"] as? String, // encryption
+            encoding == "aesgcm" {
+            plaintext = subscription.aesgcm(payload: payload, encryptionHeader: encryptionHeader, cryptoHeader: cryptoKeyHeader)
+        } else if encoding == "aes128gcm" {
+            plaintext = subscription.aes128gcm(payload: payload)
+        } else {
+            plaintext = nil
+        }
+
+        guard let _ = plaintext else {
+            return deferMaybe(PushMessageError.notDecrypted)
+        }
+
+        return handle(message: JSON(parseJSON: plaintext!))
+    }
+
+    /// The main entry point to the handler.
+    func handle(message json: JSON) -> Success {
+        if !json.isDictionary() {
+            return handleVerification()
+        }
+
+        // https://dxr.mozilla.org/mozilla-central/source/mobile/android/services/src/main/java/org/mozilla/gecko/fxa/FxAccountPushHandler.java#19
+
+
+        // https://github.com/mozilla/fxa-auth-server/blob/master/docs/pushpayloads.schema.json#L26
+
+        let rawValue = json["command"].stringValue
+        guard let command = PushMessageType(rawValue: rawValue) else {
+            log.warning("Command \(rawValue) received but not recognized")
+            return deferMaybe(PushMessageError.messageIncomplete)
+        }
+
+        let result: Success
+        switch command {
+            case .deviceConnected:
+                result = handleDeviceConnected(json["data"])
+            case .deviceDisconnected:
+                result = handleDeviceDisconnected(json["data"])
+            case .profileUpdated:
+                result = handleProfileUpdated()
+            case .passwordChanged:
+                result = handlePasswordChanged()
+            case .passwordReset:
+                result = handlePasswordReset()
+            case .collectionChanged:
+                result = handleCollectionChanged(json["data"])
+        }
+        return result
+    }
+}
+
+extension FxAPushMessageHandler {
+    func handleVerification() -> Success {
+        guard let account = profile.getAccount(), account.actionNeeded == .needsVerification else {
+            log.info("Account verified by server either doesn't exist or doesn't need verifying")
+            return deferMaybe(PushMessageError.noActionNeeded)
+        }
+
+        // Now we're verified, we can start syncing.
+        return account.advance().bind { _ in
```

Comment:
I wanted to use that before, however: `succeed` takes zero args, so you need a closure to swallow the `FxAState` parameter.

```
Cannot convert value of type '() -> Success' to expected argument type '(FxAState) -> Deferred<Maybe<Void>>'
```